core/test/solver/chebyshev.cpp

```diff
@@ -0,0 +1,373 @@
+    using value_type = typename TestFixture::value_type;
+    std::shared_ptr<Solver> chebyshev_solver =
+        Solver::build()
+            .with_criteria(
+                gko::stop::Iteration::build().with_max_iters(3u).on(this->exec))
+            .with_foci(value_type{0.2}, value_type{1.2})
+            .on(this->exec)
+            ->generate(this->mtx);
+
+    ASSERT_EQ(chebyshev_solver->get_parameters().foci,
+              std::make_pair(value_type{0.2}, value_type{1.2}));
+}
+
+
+TYPED_TEST(Chebyshev, CanSetInnerSolverInFactory)
+{
+    using Solver = typename TestFixture::Solver;
+    using value_type = typename TestFixture::value_type;
+    auto chebyshev_factory =
+        Solver::build()
+            .with_criteria(
+                gko::stop::Iteration::build().with_max_iters(3u).on(this->exec))
+            .with_preconditioner(
+                Solver::build()
+                    .with_criteria(
+                        gko::stop::Iteration::build().with_max_iters(3u).on(
+                            this->exec))
+                    .on(this->exec))
+            .on(this->exec);
+    auto solver = chebyshev_factory->generate(this->mtx);
+    auto preconditioner = dynamic_cast<const Solver*>(
+        static_cast<Solver*>(solver.get())->get_preconditioner().get());
+
+    ASSERT_NE(preconditioner, nullptr);
+    ASSERT_EQ(preconditioner->get_size(), gko::dim<2>(3, 3));
+    ASSERT_EQ(preconditioner->get_system_matrix(), this->mtx);
+}
+
+
+TYPED_TEST(Chebyshev, CanSetGeneratedInnerSolverInFactory)
+{
+    using Solver = typename TestFixture::Solver;
+    std::shared_ptr<Solver> chebyshev_solver =
+        Solver::build()
+            .with_criteria(
+                gko::stop::Iteration::build().with_max_iters(3u).on(this->exec))
+            .on(this->exec)
+            ->generate(this->mtx);
+
+    auto chebyshev_factory =
+        Solver::build()
+            .with_criteria(
+                gko::stop::Iteration::build().with_max_iters(3u).on(this->exec))
+            .with_generated_preconditioner(chebyshev_solver)
+            .on(this->exec);
+    auto solver = chebyshev_factory->generate(this->mtx);
+    auto preconditioner = solver->get_preconditioner();
+
+    ASSERT_NE(preconditioner.get(), nullptr);
+    ASSERT_EQ(preconditioner.get(), chebyshev_solver.get());
+}
+
+
+TYPED_TEST(Chebyshev, CanSetCriteriaAgain)
+{
+    using Solver = typename TestFixture::Solver;
+    std::shared_ptr<gko::stop::CriterionFactory> init_crit =
+        gko::stop::Iteration::build().with_max_iters(3u).on(this->exec);
+    auto chebyshev_factory =
+        Solver::build().with_criteria(init_crit).on(this->exec);
+
+    ASSERT_EQ((chebyshev_factory->get_parameters().criteria).back(), init_crit);
+
+    auto solver = chebyshev_factory->generate(this->mtx);
+    std::shared_ptr<gko::stop::CriterionFactory> new_crit =
+        gko::stop::Iteration::build().with_max_iters(5u).on(this->exec);
+
+    solver->set_stop_criterion_factory(new_crit);
```

Comment:
I think we already had a discussion somewhere that setting the criterion doesn't make sense. Perhaps we can start here to remove that functionality. So I would suggest to override `set_stop_criterion_factory` in the chebyshev solver and make that throw/deprecated.

Reply:
You're right. I guess we first have to put in some effort to deprecate it and show users better alternatives.